### ISSUE TITLE
[feature request] Possibility of returning only require_grad parameters in Module.parameters()

### ISSUE THREAD

**Issue by domaala** [pytorch/pytorch#7338]
Quite a few times I've encountered the situation where, instead of

```
Adam(model.parameters()), lr=1e-3)
```
I need to use
```
Adam(filter(lambda p: p.requires_grad, model.parameters()), lr=1e-3)
```
to avoid the exception of trying to optimize the parameters that do not require it. Honestly, this looks a bit ugly. I propose either including the `only_grad` flag in the `parameters()` method, or include a method `grad_parameters()`, both of which would result in a method returning only the parameters that have `requires_grad `set to `True`.

If this has been tackled before, sorry about that; I've tried searching through both closed and open issues, but I wasn't able to find what I was looking for. In case the feature idea sounds plausible, I could make a PR for it.

**Comment by zou3519** [comment]
Python list comprehensions make your code look nicer: 
```Adam([p for p in model.parameters() if p.requires_grad), lr=1e-3)```

In addition, the optimizers won't try to optimize over parameters that don't have `requires_grad` (because no gradients will accumulate in them). Do you find that this isn't the case?

**Comment by domaala** [comment]
Thanks for the reply! Even though using a list comprehension is a tad bit nicer, I still believe having to do the filtering by yourself is unnecessary.

As for your question, when I try to initialize (in this case) `Adam` without this filtering, I get a `ValueError`:
https://github.com/pytorch/pytorch/blob/4c511075c3782e96d49f16a32344a1aba3c0fd8f/torch/optim/optimizer.py#L193

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080
GPU 1: GeForce GTX 1080

Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.5.1.5
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6.0.21
/usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.14.1)
[pip3] torch (0.4.0)
[pip3] torchtext (0.2.3)
[pip3] torchvision (0.2.1)
[conda] Could not collect

**Comment by bingo619** [comment]
I think if the tensor does not require gradients at all, you should not declare it as Parameter.

If you want to freeze your subgraph, you should do it explicitly as described  on https://pytorch.org/docs/master/notes/autograd.html

**Comment by domaala** [comment]
@bingo619 Thanks for the reply!

Unfortunately, it's not that easy. It is not always possible to make the decision about having `Tensor` vs. having `Parameter`. For example, consider the `Embedding.from_pretrained()` method:

https://github.com/pytorch/pytorch/blob/f439ba58434e30d5920bae2962773b415e04c0fb/torch/nn/modules/sparse.py#L125

If you set the flag `freeze` to `True`, you will get a `Parameter` that does never require gradients, which will cause the optimizer to complain.

**Comment by apaszke** [comment]
Yeah that makes sense. There's really nothing that prevents people from setting `requires_grad` to `False` after they construct the optimizer, and it won't even complain. We should just remove the check.

**Comment by domaala** [comment]
@apaszke Agreed! Let me know whether you want me to make a PR that does that or will you handle it yourself. (Or we'll wait for more people to share their opinion on this.)

**Comment by apaszke** [comment]
Please do send a PR!

**Comment by soumith** [comment]
fixed via https://github.com/pytorch/pytorch/pull/7419

**Comment by cslxiao** [comment]
Nice
